### PR TITLE
`Development`: Refactor page util to avoid code duplication

### PR DESCRIPTION
**Pull request by rstief**
(Used to be stacked on #7903 )

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
The PageUtil.java File has a lot of code duplication

### Description
This PR reduces a lot of the code duplication in the PageUtil.java File

### Steps for Testing
Required: 1 Instructor

Test **only the sorting** of all following paginated dialogs/views:

**1. Competency Management**

- Competency Import
- Import of all Competencies of another course
- Prerequisite Import 

**2. Exercise Management**

- All Exercise Imports

**3. Exam Management**

- Exam Import
- Adding a bonus to an exam 
  -  In `3. Select Bonus Source` see that they are sorted by *Grade Key* ID descending (cannot be changed)
  - (use [this link on ts1](https://artemis-test1.artemis.cit.tum.de/course-management/16/exams/51/bonus) with test user 11 for your convenience)

**4. Other Management**

- Lecture Import
- Learning Path Management (view)

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [PageUtil.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5992/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.util/PageUtil.html) | 91% | ✅  |

## Summary by CodeRabbit

- **New Features**
  - Enhanced search functionality across various modules with column mapping for improved user experience.

- **Refactor**
  - Unified pagination and sorting logic for consistency and maintainability.

- **Bug Fixes**
  - Removed unnecessary import statements to optimize performance.

### CODE REVIEW

**Comment by rstief** [comment]
@coderabbitai review

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java [hunk 1]
```diff
@@ -1356,7 +1356,7 @@ public void scheduleModelingExercises(Exam exam) {
      * @return A wrapper object containing a list of all found exercises and the total number of pages
      */
     public SearchResultPageDTO<Exam> getAllOnPageWithSize(final PageableSearchDTO<String> search, final User user, final boolean withExercises) {
-        final var pageable = PageUtil.createExamPageRequest(search);
+        final var pageable = PageUtil.createDefaultPageRequest(search, Exam.ExamSearchColumn::getMappedColumnName);
```
[the remark this answers is not in the document]
Comment (rstief) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java]: not in the scope of this PR.

**Review by Strohgelaender — DISMISSED** [review]
Thanks for looking into this. lgtm

**Review by konrad2002 — APPROVED** [review]
manually tested on legacy ts11, works as expected

**Review by MaximilianJG — DISMISSED** [review]
Works as expected. However For the programming exercises I cannot sort by course. Is this intended? 

Screenshot 2024-01-19 at 17 20 48

**Comment by rstief** [comment]
> Works as expected. However For the programming exercises I cannot sort by course. Is this intended?

Hi Max, thanks for your tests! This is intended behaviour, to sort you would have to have only one of the `Search in Course/Exam` checkboxes checked - Then you can also search by that field 😉

**Review by MichaelOwenDyer — DISMISSED** [review]
Code looks clean and correct 👍🏽

**Review by Gusti2010 — APPROVED** [review]
Manually tested on ts2_Legacy. Sorting works as expected

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java [hunk 2]
```diff
@@ -1356,7 +1356,7 @@ public void scheduleModelingExercises(Exam exam) {
      * @return A wrapper object containing a list of all found exercises and the total number of pages
      */
     public SearchResultPageDTO<Exam> getAllOnPageWithSize(final PageableSearchDTO<String> search, final User user, final boolean withExercises) {
-        final var pageable = PageUtil.createExamPageRequest(search);
+        final var pageable = PageUtil.createDefaultPageRequest(search, PageUtil.ColumnMapping.EXAM);
         final var searchTerm = search.getSearchTerm();
         final Page<Exam> examPage;
         if (authorizationCheckService.isAdmin(user)) {
```
[the remark this answers is not in the document]
Comment (rstief) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java:1362]: not in the scope of this refactoring PR.

**Review by Strohgelaender — APPROVED** [review]
lgtm

**Review by laurenzfb — APPROVED** [review]
💯 tested on ts1

**Review by milljoniaer — APPROVED** [review]
tested in testing session, works as expected 🚀

**Review by reschandreas — APPROVED** [review]
demonstrated in testing session, works as expected. code also lgtm

**Pull request merged**